core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/dictionary/DictionaryMatcherOpExec.scala

```diff
@@ -0,0 +1,149 @@
+package edu.uci.ics.texera.workflow.operators.dictionary
+
+import edu.uci.ics.amber.engine.common.InputExhausted
+import edu.uci.ics.amber.engine.common.virtualidentity.LinkIdentity
+import edu.uci.ics.texera.workflow.common.operators.OperatorExecutor
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import edu.uci.ics.texera.workflow.common.tuple.schema.{AttributeType, OperatorSchemaInfo}
+import org.apache.lucene.analysis.Analyzer
+import org.apache.lucene.analysis.en.EnglishAnalyzer
+import org.apache.lucene.analysis.tokenattributes.OffsetAttribute
+import scala.collection.mutable.Set
+import java.io.StringReader
+import scala.collection.Iterator
+import scala.collection.mutable.ListBuffer
+import scala.util.Either
+
+class DictionaryMatcherOpExec(
+    opDesc: DictionaryMatcherOpDesc,
+    operatorSchemaInfo: OperatorSchemaInfo
+) extends OperatorExecutor {
+
+  // this is needed for the matching types Phrase and Conjunction
+  var tokenizedDictionaryEntries: ListBuffer[Set[String]] = _
+  // this is needed for the simple Scan matching type
+  var dictionaryEntries: List[String] = _
+  var luceneAnalyzer: Analyzer = _
+
+  /**
+    * first prepare the dictionary by splitting the values using a comma delimiter then tokenize the splitted values
+    */
+  override def open(): Unit = {
+    // create the dictionary by splitting the values first
+    dictionaryEntries = this.opDesc.dictionary.split(",").toList.map(_.toLowerCase)
+    // then tokenize each entry
+    this.luceneAnalyzer = new EnglishAnalyzer
+    tokenizedDictionaryEntries = ListBuffer[Set[String]]()
+    tokenizeDictionary();
+  }
+
+  override def close(): Unit = {
+    tokenizedDictionaryEntries = null
+    dictionaryEntries = null
+    luceneAnalyzer = null
+  }
+
+  /**
+    * use LuceneAnalyzer to tokenize the dictionary
+    */
+  def tokenizeDictionary(): Unit = {
+    for (text <- dictionaryEntries) {
+      tokenizedDictionaryEntries += tokenize(text)
+    }
+  }
+
+  /**
+    * decide if a tuple matches a dictionary after it is tokenized based on {@link MatchingType}
+    * @param tuple
+    * @return
+    */
+  def isTupleInDictionary(tuple: Tuple): Boolean = {
+    // first, tokenize the tuple
+    val tokenizedTuple = tokenize(tuple.getField(this.opDesc.attribute))
+    // then depending on the matching type, iterate the dictionary to match the tuple
+    if (!tokenizedTuple.isEmpty) {
+      this.opDesc.matchingType match {
+        // any subset matches regardless of their positions
+        case MatchingType.CONJUNCTION_INDEXBASED =>
+          for (entry <- tokenizedDictionaryEntries) {
+            if (entry.subsetOf(tokenizedTuple) || tokenizedTuple.subsetOf(entry)) {
```

Comment:
any reason that we are checking the subset in both directions? Isn't it enough to just check if the entry is a subset of tuple?

Reply:
to fit the user's use case of matching a url in both directions example tuple or dictionary entry is `https://www.litculture.com/product/lit-wraps-original/` should match a tuple or dictionary entry `https://www.litculture.com`. So we need both checks

Reply:
understand. if it's the user's use case, for the SCANBASED matching type, do we also want to match it both ways to be consistent?